src/sinks/datadog/logs.rs

```diff
@@ -2,23 +2,97 @@ use crate::{
     config::{log_schema, DataType, GenerateConfig, SinkConfig, SinkContext, SinkDescription},
     event::Event,
     sinks::{
-        util::{self, encoding::EncodingConfig, tcp::TcpSink, Encoding, StreamSinkOld, UriSerde},
+        util::{
+            batch::{Batch, BatchError},
+            encode_event,
+            encoding::{EncodingConfig, EncodingConfiguration},
+            http::{BatchedHttpSink, HttpClient, HttpSink},
+            BatchConfig, BatchSettings, BoxedRawValue, Compression, Encoding, JsonArrayBuffer,
+            TowerRequestConfig, VecBuffer,
+        },
         Healthcheck, VectorSink,
     },
     tls::{MaybeTlsSettings, TlsConfig},
 };
 use bytes::Bytes;
-use futures01::{stream::iter_ok, Sink};
+use flate2::write::GzEncoder;
+use futures::FutureExt;
+use futures01::Sink;
+use http::{Request, StatusCode};
+use hyper::body::Body;
 use serde::{Deserialize, Serialize};
+use serde_json::json;
+use std::{convert::TryFrom, io::Write, time::Duration};
 use string_cache::DefaultAtom as Atom;
 
-#[derive(Deserialize, Serialize, Debug)]
+/// GZip compression level must be between 0 and 9.
+/// 0 is no compression, 9 is take as long as you like for maximum compression.
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(try_from = "u32")]
+#[serde(into = "u32")]
+pub struct CompressionLevel(flate2::Compression);
+
+impl CompressionLevel {
+    fn get_level(&self) -> flate2::Compression {
+        self.0
+    }
+}
+
+impl Into<u32> for CompressionLevel {
+    fn into(self) -> u32 {
+        self.0.level()
+    }
+}
+
+impl TryFrom<u32> for CompressionLevel {
+    type Error = String;
+
+    fn try_from(level: u32) -> Result<Self, Self::Error> {
+        if level > 9 {
+            Err("Gzip `compression_level` must be between 0 and 9.".into())
+        } else {
+            Ok(CompressionLevel(flate2::Compression::new(level)))
+        }
+    }
+}
+
+impl Default for CompressionLevel {
+    fn default() -> Self {
+        // Default the compression level to 6, which is the same as the Datadog agent.
+        // https://docs.datadoghq.com/agent/logs/log_transport/?tab=https#log-compression
+        CompressionLevel(flate2::Compression::new(6))
+    }
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
 #[serde(deny_unknown_fields)]
 pub struct DatadogLogsConfig {
-    endpoint: Option<UriSerde>,
+    endpoint: Option<String>,
     api_key: String,
     encoding: EncodingConfig<Encoding>,
     tls: Option<TlsConfig>,
+
+    #[serde(default)]
+    compression: Option<Compression>,
+
+    #[serde(default)]
+    compression_level: CompressionLevel,
```

Comment:
Can we remove `compression_level` field? Compression level should be added to Compression struct in https://github.com/timberio/vector/pull/3032.

Reply:
Ah nice. I hadn't seen that pull. That's a much better way to do it. How soon is that likely to be merged in?

Reply:
I hope that we will merge it on this week :slightly_smiling_face:

Reply:
Ok great. I'll hold off on this change until that lands.

Reply:
#3032 was merged, so this can proceed.

Reply:
All done.